extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -107,28 +110,109 @@ public Uni<SecurityIdentity> apply(HttpAuthenticationMechanism mech) {
+                                                        return mech.authenticate(routingContext, identityProviderManager);
+                                                    }
+                                                });
+                            }
+
+                        }
+                        return Uni.createFrom().optional(Optional.empty());
+                    }
+                });
+            }
+
+            // Now make sure the first SecurityIdentity is returned since the current SecurityIdentity was created
+            // by the last authentication mechanism in the list
+            return result.onItem().transformToUni(new Function<SecurityIdentity, Uni<? extends SecurityIdentity>>() {
                 @Override
-                public Uni<SecurityIdentity> apply(SecurityIdentity data) {
-                    if (data != null) {
-                        return Uni.createFrom().item(data);
+                public Uni<SecurityIdentity> apply(final SecurityIdentity identity) {
+                    if (identity != null) {
+                        final SecurityIdentity firstIdentity = routingContext
+                                .get(HttpSecurityUtils.SECURITY_IDENTITY_ATTRIBUTE);
+                        if (firstIdentity != null) {
+                            final Map<String, SecurityIdentity> identities = getSecurityIdentities(firstIdentity);
+
+                            return getCredentialTransport(mechanisms[mechanisms.length - 1], routingContext)
+                                    .onItem().transform(new Function<HttpCredentialTransport, SecurityIdentity>() {
+                                        @Override
+                                        public SecurityIdentity apply(HttpCredentialTransport transport) {
+                                            // Save the extra identity in the identities map
+                                            identities.put(transport.getAuthenticationScheme(), identity);
+                                            // Return the first identity
+                                            return firstIdentity;
+                                        }
+                                    });
+                        } else {
+                            // Single authentication mechanism is available only
+                            return Uni.createFrom().item(identity);
+                        }
                     }
-                    return mech.authenticate(routingContext, identityProviderManager);
+                    return Uni.createFrom().optional(Optional.empty());
                 }
             });
         }
-        return result;
+    }
+
+    private Map<String, SecurityIdentity> getSecurityIdentities(SecurityIdentity firstIdentity) {
+        Map<String, SecurityIdentity> identities = firstIdentity.getAttribute(HttpSecurityUtils.SECURITY_IDENTITIES_ATTRIBUTE);
```

Comment:
I really don't like the way this is implemented, as it forces a previously immutable identity to be mutable. 

IMHO I think you should consider chaining them the other way, so the last identity is the one exposed to the application. This way you could keep them immutable (although you will need to use chaining rather than combine like I mention above). This last identity has an attribute that points to the previous identity in the chain.

Reply:
Note that you can still get the same results (i.e. the mTLS identity being injected), just reverse the priority order.

Reply:
@stuartwdouglas  Stuart, but even if we do it this way, how thie current identity can be updated to link to the previous one with an extra attribute?  Individual identities are created by authentication mechanisms, they don't know about other identities,  while HttpAuthenticator  does not have access to the identity builder ?

See what I mean ? This is what the workaround in the PR tries to achieve,  allow HttpAuthenticator add the attribute,  I don't mind combining Unis to simplify, and just having the current identity linking to the previous one but the challenge is how to add this link 😀

Reply:
I think you just have to wrap them, the same way SecurityIdentityAugmentor already needs to.